File: ios/headmaster/LivenessModule.swift
import Foundation
import UIKit
import Vision
import React

@objc(LivenessModule)
class LivenessModule: NSObject {

  @objc static func requiresMainQueueSetup() -> Bool { false }

  @objc func detectBlinkFromFrames(
    _ imagePaths: [String],
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard imagePaths.count >= 3 else {
      reject("INSUFFICIENT_FRAMES", "Need at least 3 frames", nil)
      return
    }

    var eyeOpenValues: [Float] = []
    let group = DispatchGroup()

    for path in imagePaths {
      group.enter()
      guard let image = loadImage(from: path) else {
        group.leave()
        continue
      }

      detectFace(in: image) { face in
        if let face = face {
          eyeOpenValues.append(face)
        }
        group.leave()
      }
    }

    group.notify(queue: .main) {
      guard eyeOpenValues.count >= 3 else {
        reject("NO_FACE", "Could not detect face. Hold still and blink.", nil)
        return
      }

      var blinkDetected = false
      for i in 1..<eyeOpenValues.count {
        let drop = eyeOpenValues[i - 1] - eyeOpenValues[i]
        if drop > 0.3 && eyeOpenValues[i] < 0.4 {
          blinkDetected = true
          break
        }
      }

      if !blinkDetected {
        let minVal = eyeOpenValues.min() ?? 1
        let maxVal = eyeOpenValues.max() ?? 0
        if maxVal - minVal > 0.35 && minVal < 0.35 {
          blinkDetected = true
        }
      }

      resolve([
        "blinkDetected": blinkDetected,
        "faceDetected": true,
        "confidence": blinkDetected ? 92.0 : 30.0,
        "message": blinkDetected
          ? "Liveness verified successfully"
          : "Please blink your eyes naturally and try again"
      ])
    }
  }

  @objc func validateFaceQuality(
    _ imagePath: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let image = loadImage(from: imagePath) else {
      reject("FILE_NOT_FOUND", "Image not found", nil)
      return
    }

    detectFaceDetails(in: image) { result in
      resolve(result)
    }
  }

  private func loadImage(from path: String) -> UIImage? {
    let cleanPath = path.replacingOccurrences(of: "file://", with: "")
    if let image = UIImage(contentsOfFile: cleanPath) {
      return image
    }
    if let url = URL(string: path), let data = try? Data(contentsOf: url) {
      return UIImage(data: data)
    }
    return nil
  }

  private func detectFace(in image: UIImage, completion: @escaping (Float?) -> Void) {
    guard let cgImage = image.cgImage else {
      completion(nil)
      return
    }

    let request = VNDetectFaceLandmarksRequest { request, _ in
      guard let results = request.results as? [VNFaceObservation],
            let face = results.first,
            let landmarks = face.landmarks,
            let leftEye = landmarks.leftEye,
            let rightEye = landmarks.rightEye else {
        completion(nil)
        return
      }

      let leftOpen = self.estimateEyeOpenness(leftEye)
      let rightOpen = self.estimateEyeOpenness(rightEye)
      completion((leftOpen + rightOpen) / 2)
    }

    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    DispatchQueue.global(qos: .userInitiated).async {
      try? handler.perform([request])
    }
  }

  private func detectFaceDetails(in image: UIImage, completion: @escaping ([String: Any]) -> Void) {
    guard let cgImage = image.cgImage else {
      completion(["valid": false, "message": "Invalid image"])
      return
    }

    let request = VNDetectFaceLandmarksRequest { request, _ in
      guard let results = request.results as? [VNFaceObservation],
            let face = results.first else {
        completion(["valid": false, "message": "No face detected. Center your face."])
        return
      }

      let headY = abs(face.yaw?.doubleValue ?? 0) * 180 / .pi
      let headZ = abs(face.roll?.doubleValue ?? 0) * 180 / .pi
      let leftOpen = face.landmarks?.leftEye.map { self.estimateEyeOpenness($0) } ?? 0.5
      let rightOpen = face.landmarks?.rightEye.map { self.estimateEyeOpenness($0) } ?? 0.5

      let valid = leftOpen > 0.5 && rightOpen > 0.5 && headY < 20 && headZ < 15
      var message = "Face quality is good"
      if headY >= 20 { message = "Turn your face straight to the camera" }
      else if headZ >= 15 { message = "Keep your head level" }
      else if leftOpen <= 0.5 || rightOpen <= 0.5 { message = "Open your eyes fully" }

      completion([
        "valid": valid,
        "leftEyeOpen": leftOpen,
        "rightEyeOpen": rightOpen,
        "headTurnY": headY,
        "headTurnZ": headZ,
        "message": message
      ])
    }

    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    DispatchQueue.global(qos: .userInitiated).async {
      try? handler.perform([request])
    }
  }

  private func estimateEyeOpenness(_ eye: VNFaceLandmarkRegion2D) -> Float {
    let points = eye.normalizedPoints
    guard points.count >= 4 else { return 0.5 }
    let minY = points.map { $0.y }.min() ?? 0
    let maxY = points.map { $0.y }.max() ?? 0
    let openness = Float(maxY - minY) * 10
    return min(1, max(0, openness))
  }
}

File: android/app/src/main/java/com/headmaster/liveness/LivenessModule.kt
package com.headmaster.liveness

import android.graphics.BitmapFactory
import android.net.Uri
import com.facebook.react.bridge.*
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetector
import com.google.mlkit.vision.face.FaceDetectorOptions
import java.io.File
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import kotlin.math.abs

class LivenessModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "LivenessModule"

    private fun buildDetector(): FaceDetector {
        val options = FaceDetectorOptions.Builder()
            .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
            .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
            .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
            .setMinFaceSize(0.15f)
            .enableTracking()
            .build()
        return FaceDetection.getClient(options)
    }

    @ReactMethod
    fun detectBlinkFromFrames(imagePaths: ReadableArray, promise: Promise) {
        if (imagePaths.size() < 3) {
            promise.reject("INSUFFICIENT_FRAMES", "Need at least 3 frames for blink detection")
            return
        }

        val detector = buildDetector()
        val eyeStates = mutableListOf<Pair<Float, Float>>()
        var processed = 0
        val latch = CountDownLatch(imagePaths.size())

        for (i in 0 until imagePaths.size()) {
            val path = imagePaths.getString(i) ?: continue
            val file = File(path.replace("file://", ""))
            if (!file.exists()) {
                latch.countDown()
                continue
            }

            val bitmap = BitmapFactory.decodeFile(file.absolutePath)
            if (bitmap == null) {
                latch.countDown()
                continue
            }

            val image = InputImage.fromBitmap(bitmap, 0)
            detector.process(image)
                .addOnSuccessListener { faces ->
                    if (faces.isNotEmpty()) {
                        val face = faces[0]
                        val left = face.leftEyeOpenProbability ?: 0.5f
                        val right = face.rightEyeOpenProbability ?: 0.5f
                        synchronized(eyeStates) {
                            eyeStates.add(Pair(left, right))
                        }
                    }
                    processed++
                    latch.countDown()
                }
                .addOnFailureListener {
                    latch.countDown()
                }
        }

        Thread {
            try {
                latch.await(10, TimeUnit.SECONDS)
                detector.close()

                if (eyeStates.size < 3) {
                    promise.reject("NO_FACE", "Could not detect face in enough frames. Hold still and blink.")
                    return@Thread
                }

                val avgOpen = eyeStates.map { (l, r) -> (l + r) / 2f }
                var blinkDetected = false

                for (i in 1 until avgOpen.size) {
                    val drop = avgOpen[i - 1] - avgOpen[i]
                    // Soft blink threshold — works with lighting lag on phones
                    if (drop > 0.12f && avgOpen[i] < 0.6f) {
                        blinkDetected = true
                        break
                    }
                    // open after closed
                    val rise = avgOpen[i] - avgOpen[i - 1]
                    if (rise > 0.12f && avgOpen[i - 1] < 0.55f) {
                        blinkDetected = true
                        break
                    }
                }

                // Eyes closed then open range pattern
                if (!blinkDetected) {
                    val minVal = avgOpen.minOrNull() ?: 1f
                    val maxVal = avgOpen.maxOrNull() ?: 0f
                    if (maxVal - minVal > 0.15f && minVal < 0.55f) {
                        blinkDetected = true
                    }
                }

                // Soft fallback: face across frames + mild eye variation (slow blinks)
                if (!blinkDetected && eyeStates.size >= 4) {
                    val variation = (avgOpen.maxOrNull() ?: 0f) - (avgOpen.minOrNull() ?: 0f)
                    if (variation > 0.08f) {
                        blinkDetected = true
                    }
                }

                val result = Arguments.createMap().apply {
                    putBoolean("blinkDetected", blinkDetected)
                    putBoolean("faceDetected", true)
                    putDouble("confidence", if (blinkDetected) 92.0 else 30.0)
                    putString("message", if (blinkDetected)
                        "Liveness verified successfully"
                    else
                        "Please blink your eyes naturally and try again")
                }
                promise.resolve(result)
            } catch (e: Exception) {
                promise.reject("LIVENESS_ERROR", e.message, e)
            }
        }.start()
    }

    @ReactMethod
    fun validateFaceQuality(imagePath: String, promise: Promise) {
        val detector = buildDetector()
        val file = File(imagePath.replace("file://", ""))

        if (!file.exists()) {
            promise.reject("FILE_NOT_FOUND", "Image file not found")
            return
        }

        val bitmap = BitmapFactory.decodeFile(file.absolutePath)
        if (bitmap == null) {
            promise.reject("INVALID_IMAGE", "Could not decode image")
            return
        }

        val image = InputImage.fromBitmap(bitmap, 0)
        detector.process(image)
            .addOnSuccessListener { faces ->
                detector.close()
                if (faces.isEmpty()) {
                    val result = Arguments.createMap().apply {
                        putBoolean("valid", false)
                        putString("message", "No face detected. Center your face in the frame.")
                    }
                    promise.resolve(result)
                    return@addOnSuccessListener
                }

                val face = faces[0]
                val leftEye = face.leftEyeOpenProbability ?: 0.5f
                val rightEye = face.rightEyeOpenProbability ?: 0.5f
                val smiling = face.smilingProbability ?: 0f
                val headEulerY = abs(face.headEulerAngleY)
                val headEulerZ = abs(face.headEulerAngleZ)

                val valid = leftEye > 0.5f && rightEye > 0.5f &&
                    headEulerY < 20f && headEulerZ < 15f

                val result = Arguments.createMap().apply {
                    putBoolean("valid", valid)
                    putDouble("leftEyeOpen", leftEye.toDouble())
                    putDouble("rightEyeOpen", rightEye.toDouble())
                    putDouble("smilingProbability", smiling.toDouble())
                    putDouble("headTurnY", headEulerY.toDouble())
                    putDouble("headTurnZ", headEulerZ.toDouble())
                    putString("message", when {
                        headEulerY >= 20f -> "Turn your face straight to the camera"
                        headEulerZ >= 15f -> "Keep your head level"
                        leftEye <= 0.5f || rightEye <= 0.5f -> "Open your eyes fully"
                        else -> "Face quality is good"
                    })
                }
                promise.resolve(result)
            }
            .addOnFailureListener { e ->
                detector.close()
                promise.reject("FACE_ERROR", e.message, e)
            }
    }
}
